File: src/Ycs/Core/ContentFactory.cs
// ------------------------------------------------------------------------------
//  <copyright company="Microsoft Corporation">
//      Copyright (c) Microsoft Corporation.  All rights reserved.
//  </copyright>
// ------------------------------------------------------------------------------

using System;
using Ycs.Contracts;
using Ycs.Content;
using Ycs.Types;

namespace Ycs.Core
{
    /// <summary>
    /// Default implementation of content factory
    /// </summary>
    public class ContentFactory : IContentFactory
    {
        public IContent CreateContentType(IAbstractType type)
        {
            return new ContentType(type);
        }
        
        public IContent CreateContentDoc(object doc)
        {
            if (doc is YDoc yDoc)
            {
                return new ContentDoc(yDoc);
            }
            throw new ArgumentException("Expected YDoc instance", nameof(doc));
        }

        public IContent CreateContentBinary(byte[] ba)
        {
            return new ContentBinary(ba);
        }

        public IContent CreateContentAny(object value)
        {
            return new ContentAny(new object[] { value });
        }

        public IContent CreateContentFormat(string key, object value)
        {
            return new ContentFormat(key, value);
        }

        public IContent CreateContentString(string text)
        {
            return new ContentString(text);
        }

        public IContent CreateContentEmbed(object embed)
        {
            return new ContentEmbed(embed);
        }


        public IContent CreateContent(object value)
        {
            if (value == null)
            {
                return CreateContentAny(value);
            }
            switch (value)
            {
                case YDoc d:
                    return CreateContentDoc(d);
                case AbstractType at:
                    return CreateContentType(at);
                case byte[] ba:
                    return CreateContentBinary(ba);
                default:
                    return CreateContentAny(value);
            }
        }
        
 
    }
}

File: ycs-golang/core/content_factory.go
// ------------------------------------------------------------------------------
//  <copyright company="Microsoft Corporation">
//      Copyright (c) Microsoft Corporation.  All rights reserved.
//  </copyright>
// ------------------------------------------------------------------------------

package core

import (
	"fmt"

	"github.com/chenrensong/ygo/content"
	"github.com/chenrensong/ygo/contracts"
	"github.com/chenrensong/ygo/types"
)

// ContentFactory is the default implementation of content factory
type ContentFactory struct{}

// NewContentFactory creates a new ContentFactory
func NewContentFactory() contracts.IContentFactory {
	return &ContentFactory{}
}

// CreateContentType creates a ContentType
func (cf *ContentFactory) CreateContentType(abstractType contracts.IAbstractType) contracts.IContent {
	return content.NewContentType(abstractType)
}

// CreateContentDoc creates a ContentDoc
func (cf *ContentFactory) CreateContentDoc(doc interface{}) contracts.IContent {
	if yDoc, ok := doc.(*types.YDoc); ok {
		return content.NewContentDoc(yDoc)
	}
	panic(fmt.Sprintf("Expected YDoc instance, got %T", doc))
}

// CreateContentBinary creates a ContentBinary
func (cf *ContentFactory) CreateContentBinary(ba []byte) contracts.IContent {
	return content.NewContentBinary(ba)
}

// CreateContentAny creates a ContentAny
func (cf *ContentFactory) CreateContentAny(value interface{}) contracts.IContent {
	return content.NewContentAny([]interface{}{value})
}

// CreateContentFormat creates a ContentFormat
func (cf *ContentFactory) CreateContentFormat(key string, value interface{}) contracts.IContent {
	return content.NewContentFormat(key, value)
}

// CreateContentString creates a ContentString
func (cf *ContentFactory) CreateContentString(text string) contracts.IContent {
	return content.NewContentString(text)
}

// CreateContentEmbed creates a ContentEmbed
func (cf *ContentFactory) CreateContentEmbed(embed interface{}) contracts.IContent {
	return content.NewContentEmbed(embed)
}

// CreateContent creates content based on the value type
func (cf *ContentFactory) CreateContent(value interface{}) contracts.IContent {
	if value == nil {
		return cf.CreateContentAny(value)
	}

	switch v := value.(type) {
	case *types.YDoc:
		return cf.CreateContentDoc(v)
	case contracts.IAbstractType:
		return cf.CreateContentType(v)
	case []byte:
		return cf.CreateContentBinary(v)
	default:
		return cf.CreateContentAny(value)
	}
}
